res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,719 @@
+//     * changed function Shift + PAD in Sampler Mode
+//         - while playing Sample stops play and jump to start
+//         - while not playing load selected track into PAD
+//     * added Keyshift Mode (Sampler +Shift)
+//     * added BROWSE +Shift - Waveform zoom +/-
+//
+// Version 0.2 - 15.05.2019:
+//     * added Beatjump Beatcount manipulation increase (pad 8 + shift), decrease (pad 7 +shift)
+//     * fixed a bug in beatloop with 1/4 and 1/2 Beat loops
+//     * fixed a bug in BEAT SYNC + Shift (cycle tempo range)
+//     * added support for saving and playing loops on Hot Cue Pads
+//     * added support for loop in + out adjust (while looping press and hold in or out + jog wheel to adjust the loop poit position)
+//     * fixed a bug in Tempo Range switch (Beat Sync + shift)
+//
+// Version 0.3 - 18.05.2019:
+//     * added Effect Section funtions
+//         - Beat FX left and right selects the Effect Slot in FX3
+//         - Shift + Beat FX On/Off disables all Effect Slots
+//
+// version 0.4:
+//     * added Cue/Loop left and richt to navigate thru Loop / hotcue points
+//     * initial support for some LEDs (VuMeter, Play, Cue, Loop in/out, sync)
+//
+var PioneerDDJ400 = {};
+
+var LightsPioneerDDJ400 = {
+    beatFx: {
+        status: 0x94,
+        data1: 0x47,
+    },
+    deck1: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x90,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x90,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x90,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x90,
+            data1: 0x48,
+        },
+    },
+    deck2: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x91,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x91,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x91,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x91,
+            data1: 0x48,
+        },
+    },
+};
+
+// Save the Shift State
+PioneerDDJ400.shiftState = [0, 0];
+
+// JogWheel
+PioneerDDJ400.vinylMode = true;
+PioneerDDJ400.alpha = 1.0 / 8;
+PioneerDDJ400.beta = PioneerDDJ400.alpha / 32;
+PioneerDDJ400.highspeedScale = 2;
+PioneerDDJ400.bendScale = 0.5;
+
+PioneerDDJ400.pointJumpSpace = 0.005; // amount in percent of the Song we can jump back to previous Cue or loop point
+
+PioneerDDJ400.tempoRanges = [0.06, 0.10, 0.16, 0.25]; // WIDE = 25%?
+
+// Keyboard Mode Variables and Settings
+PioneerDDJ400.keyboardHotCuePoint = [0, 0]; // selected HotCue point (eg. PAD) in Keyboard mode per Deck 0 = unset
+PioneerDDJ400.keyboardModeRefCount = [0, 0]; // count the currently pressed Pads per Deck
+PioneerDDJ400.halftoneToPadMap = [4, 5, 6, 7, 0, 1, 2, 3];
```

Comment:
This does not make sense to me. If you have 8 pads, I think the best way to use them would be to map one octave of a major scale.